### DISCUSSION TITLE
Why does ClusterClient.TxPipeline execute slots cmds squentially，but not concurrentlly？

### DISCUSSION THREAD

**Discussion by YangXin-Sheep** [redis/go-redis#3200] [Q&A]
every slot is independence，so I think different slot can execute cms concurrentlly.

**Comment by Pleo2** [comment]
The important distinction is that `TxPipeline` is not merely a faster pipeline: it asks the client to wrap commands in `MULTI`/`EXEC`.

In go-redis v9.7.0, commands were grouped by hash slot and the outer slot loop was sequential ([source](https://github.com/redis/go-redis/blob/v9.7.0/osscluster.go#L1428-L1490)). Each slot therefore received a separate transaction; the whole call was **not** one atomic cross-slot transaction. Running those groups concurrently could improve latency, but it would still be multiple independent transactions and could create a misleading atomicity expectation.

The current behavior makes that boundary explicit: `ClusterClient.TxPipeline` rejects commands spanning more than one slot with `ErrCrossSlot` ([v9.22.0 source](https://github.com/redis/go-redis/blob/v9.22.0/osscluster.go#L2261-L2274)).

So the practical choices are:

- Use `Pipeline`/`Pipelined` when operations on different slots are independent and you want batching/concurrency without transaction semantics.
- Use `TxPipeline` only when every keyed command is intentionally in one slot. Redis hash tags such as `account:{42}:balance` and `account:{42}:ledger` can co-locate related keys.

That preserves the semantic difference: concurrency across independent shards belongs to a normal pipeline; atomic `MULTI`/`EXEC` belongs to one shard/slot.
